tools/code-analyzer/src/git.ts

```diff
@@ -5,12 +5,135 @@ import { CliUx } from '@oclif/core';
 import { execSync } from 'child_process';
 import { join } from 'path';
 import { tmpdir } from 'os';
-import { readFileSync } from 'fs';
+import { readFileSync, rmSync } from 'fs';
+import { simpleGit } from 'simple-git';
+import { v4 } from 'uuid';
 
 /**
  * Internal dependencies
  */
 import { startWPEnv, stopWPEnv, isValidCommitHash } from './utils';
+/**
+ * Clone a git repository.
+ *
+ * @param {string} repoPath- the path (either URL or file path) to the repo to clone.
+ * @param          repoPath
+ * @return {string} the path to the cloned repo.
+ */
+export const cloneRepo = async ( repoPath: string ) => {
+	const tempFolderName = v4();
+	const git = simpleGit( { baseDir: tmpdir() } );
+	await git.clone( repoPath, tempFolderName );
+
+	return tempFolderName;
+};
+
+/**
+ * Do a git diff of 2 commit hashes (or branches)
+ *
+ * @param {string} baseDir - baseDir that the repo is in
+ * @param {string} hashA   - either a git commit hash or a git branch
+ * @param {string} hashB   - either a git commit hash or a git branch
+ * @return {string} - diff of the changfiles between the 2 hashes
+ */
+export const diffHashes = ( baseDir: string, hashA: string, hashB: string ) => {
+	const git = simpleGit( { baseDir } );
+	return git.diff( [ `${ hashA }..${ hashB }` ] );
+};
+
+/**
+ * Determines if a string is a commit hash or not.
+ *
+ * @param {string} ref - the ref to check
+ * @return {boolean} whether the ref is a commit hash or not.
+ */
+const refIsHash = ( ref: string ) => {
+	return /^[0-9a-f]{7,40}$/i.test( ref );
+};
+
+/**
+ * Get the commit hash for a ref (either branch or commit hash). If a validly
+ * formed hash is provided it is returned unmodified.
+ *
+ * @param {string} baseDir - the dir of the git repo to get the hash from.
+ * @param {string} ref     -    Either a commit hash or a branch name.
+ * @return {string} - the commit hash of the ref.
+ */
+export const getCommitHash = async ( baseDir: string, ref: string ) => {
+	const isHash = refIsHash( ref );
+
+	// If its not a hash we assume its a branch
+	if ( ! isHash ) {
+		return simpleGit( { baseDir } ).revparse( [ ref ] );
+	}
+
+	// Its a hash already
+	return ref;
+};
+
+/**
+ * generateDiff generates a diff for a given repo and 2 hashes or branch names.
+ *
+ * @param {string}   repoPath - the url or filepath of the repo to clone.
+ * @param {string}   hashA    - commit hash or branch name.
+ * @param {string}   hashB    - commit hash or branch name.
+ * @param {Function} onError  - the handler to call when an error occurs.
+ */
+export const generateDiff = async (
+	repoPath: string,
+	hashA: string,
+	hashB: string,
+	onError: ( error: string ) => void
+) => {
+	try {
+		const context = await cloneRepo( repoPath );
+		const tmpRepoPath = join( tmpdir(), context );
```

Comment:
Ah I could improve this. `cloneRepo` just returns the uuid generated for the folder name at the moment but you're right that `tmpdir` is called in there so I could return the fully joined path instead. I'll fix that.

Reply:
Did some major refactoring of this to solve other issues as well. Most of it can be seen in https://github.com/woocommerce/woocommerce/pull/33837/commits/fe05294d9d3e3caa2e51d876731f1a5debd961b0